src/Aave.sol

```diff
@@ -0,0 +1,249 @@
+// bytes are provided, the max amount is used. If one byte is provided, that
+// data is considered the numerator, and the denominator becomes 255, which is
+// the max value of a single byte. If two bytes are provided, the data is still
+// considered the numerator, but the denominator becomes 65_535, which is the
+// max value of 2 bytes. This pattern continues through 31 bytes, and this
+// method will revert if you try passing 32 bytes or more of calldata.
+// Realistically you'll never need that much anyway.
+function parseAmount(uint balance, bytes calldata data) pure returns (uint) {
+  if (data.length == 0) {
+    return balance;
+  }
+
+  uint bits = data.length * 8;
+  uint fraction = uint(bytes32(data) >> (256 - bits));
+  uint maxUintN = (1 << bits) - 1;
+  return balance * fraction / maxUintN;
+}
+
+// =========================
+// ======== Routers ========
+// =========================
+
+abstract contract AaveRouterBase {
+  IAavePool public immutable AAVE;
+  address public immutable ASSET;
+  address public immutable ATOKEN;
+
+  uint internal constant MAX_UINT = type(uint).max;
+  uint16 internal constant REFERRAL_CODE = 0;
+
+  constructor(IAavePool aave, address asset) {
+    AAVE = aave;
+    ASSET = asset;
+    ATOKEN = aave.getReserveData(asset).aTokenAddress;
+  }
+}
+
+// -------- Supply ETH Router --------
+contract AaveSupplyEth is AaveRouterBase {
+  using SafeTransferLib for IERC20;
+
+  constructor(IAavePool aave, address asset) AaveRouterBase(aave, asset) {
+    IERC20(asset).safeApprove(address(aave), type(uint).max);
+  }
+
+  receive() external payable {
+    IWETH(ASSET).deposit{value: msg.value}();
+    AAVE.supply(ASSET, msg.value, msg.sender, REFERRAL_CODE);
+  }
+}
+
+// -------- Supply Tokens Router --------
+contract AaveSupplyToken is AaveRouterBase {
+  using SafeTransferLib for IERC20;
+
+  constructor(IAavePool aave, address asset) AaveRouterBase(aave, asset) {
+    IERC20(asset).safeApprove(address(aave), type(uint).max);
+  }
+
+  fallback() external {
+    uint balance = IERC20(ASSET).balanceOf(msg.sender);
+    uint amt = parseAmount(balance, msg.data);
+    IERC20(ASSET).safeTransferFrom(msg.sender, address(this), amt);
+    AAVE.supply(ASSET, amt, msg.sender, 0);
+  }
+}
+
+// -------- Withdraw ETH Router --------
+contract AaveWithdrawEth is AaveRouterBase {
+  using SafeTransferLib for IERC20;
+
+  constructor(IAavePool aave, address asset) AaveRouterBase(aave, asset) {}
+
+  fallback() external payable {
+    if (msg.sender == ASSET) {
+      return; // Getting ETH from the WETH contract.
+    }
+
+    uint balance = IERC20(ATOKEN).balanceOf(msg.sender);
+    uint amt = parseAmount(balance, msg.data);
+    IERC20(ATOKEN).safeTransferFrom(msg.sender, address(this), amt);
+
+    // Need to pass MAX_UINT to withdraw full amount after interest accrual.
+    amt = AAVE.withdraw(ASSET, amt == balance ? MAX_UINT : amt, address(this));
```

Comment:
It'll revert if it doesn't succeed so I don't think we need any other checks.

Also which amounts do we need to make sure match?

Reply:
> Also which amounts do we need to make sure match?

The amount we calculate from the `msg.data` vs the amount we get as a return value from `AAVE.withdraw`

Reply:
Hmm I think this is probably, since I'm fairly confident from looking at the code that it'll revert if we don't get the provided amount back. The reason for using the return value at all is just to replace `MAX_UINT` with the actual amount in those cases (since we won't know the max amount beforehand, since it depends on how much interest accrues before the withdraw)